### PR TITLE
cp-291: Sign out user on 401 error

### PR DESCRIPTION
**Pull request by OlesiaTr**
Redirection to /sign-in on 401 error occurrence

https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/assets/104642736/6207210f-b6aa-4af0-a9c9-67de3c823758

### CODE REVIEW

File: frontend/src/libs/packages/api/base-http-api.ts [hunk 1]
```diff
@@ -106,6 +110,11 @@ class BaseHttpApi implements HTTPApi {
       };
     }
 
+    if (response.status === HTTPCode.UNAUTHORIZED) {
+      await this.storage.drop(StorageKey.TOKEN);
+      window.location.assign(AppRoute.SIGN_IN);
```
Comment (v0ldemar01) [inline frontend/src/libs/packages/api/base-http-api.ts]: Not sure that is needed for us

**Review by v0ldemar01 — CHANGES_REQUESTED** [review]

File: frontend/src/libs/packages/api/base-http-api.ts (hunk 1 above)
Comment (OlesiaTr) [reply frontend/src/libs/packages/api/base-http-api.ts]: @v0ldemar01  Where do you want me to sign out user then?🤔

Comment (liza-veis) [reply frontend/src/libs/packages/api/base-http-api.ts]: we need to handle this in handleError middleware

Comment (OlesiaTr) [reply frontend/src/libs/packages/api/base-http-api.ts]: Maybe I am getting smth wrong, but I can't get error code from actions in the handleError middleware. It's working with signOut and this way error msg pop ups too, but thought that comparing with 401 would be correct

File: frontend/src/libs/packages/store/middlewares/middlewares.ts [hunk 2]
```diff
@@ -1,16 +1,22 @@
...
 import { appActions } from '#slices/app/app-notification.js';
+import { signOut } from '#slices/auth/actions.js';
 
 const handleError: Middleware<unknown, unknown, AppDispatch> = ({
   dispatch,
 }) => {
   return (next) => {
     return (action: AnyAction) => {
       if (isRejected(action)) {
+        if (action.error.message === ExceptionMessage.UNAUTHORIZED_USER) {
```
Comment (v0ldemar01) [inline frontend/src/libs/packages/store/middlewares/middlewares.ts]: We need to listen on 401 in BaseHttpApi and throw custom error and catch it here

Comment (v0ldemar01) [reply frontend/src/libs/packages/store/middlewares/middlewares.ts]: Also please create handleUnauthorized for decompose error handling

**Review by v0ldemar01 — CHANGES_REQUESTED** [review]

File: frontend/src/libs/packages/store/middlewares/middlewares.ts (hunk 2 above)
Comment (OlesiaTr) [reply frontend/src/libs/packages/store/middlewares/middlewares.ts]: @v0ldemar01 clarify for me please, so I understand it correctly.: 
1. I need to throw an HttpError in the checkResponse or handleError of BaseHttpApi? 
2. Then I need to create handle-unauthorized.middleware.ts for example? 
3. If yes, do I need to create separate file for handleError middleware and make re-exports inside of the middleware.ts? And then add them inside of the store package?

File: frontend/src/libs/packages/store/middlewares/middlewares.ts [hunk 3]
```diff
@@ -1,16 +1,22 @@
 import { type AnyAction, type Middleware } from '@reduxjs/toolkit';
 import { isRejected } from '@reduxjs/toolkit';
+import { ExceptionMessage } from 'shared/build/index.js';
```
Comment (dmytro-kucherenko) [inline frontend/src/libs/packages/store/middlewares/middlewares.ts]: I'm not sure that import from `shared` folder is allowed to be here, only via reexports

File: frontend/src/libs/packages/store/middlewares/middlewares.ts (hunk 2 above)
Comment (v0ldemar01) [reply frontend/src/libs/packages/store/middlewares/middlewares.ts]: It's not a HttpError, a custom one that extends from Error, as for handle-unauthorized.middleware.ts you are right, just Implement as you see fit

File: frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+import { ExceptionMessage } from '../../libs/enums/enums.js';
+
+const handleUnauthorized: Middleware<unknown, unknown, AppDispatch> = ({
+  dispatch,
+}) => {
+  return (next) => {
+    return (action: AnyAction) => {
+      if (isRejected(action)) {
+        const isUnauthorized =
```
Comment (liza-veis) [inline frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts]: compare them by error name, not exception message

File: frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+          action.error.message === ExceptionMessage.UNAUTHORIZED_USER;
+
+        isUnauthorized && void dispatch(authActions.signOut());
```
Comment (liza-veis) [inline frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts]:
```suggestion
       if (isUnauthorized) {
         void dispatch(authActions.signOut());
       }
```

**Review by liza-veis — CHANGES_REQUESTED** [review]

File: frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts (hunk 4 above)
Comment (OlesiaTr) [reply frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts]: @Liza-Veis  as far as I tried to test it with other errors - the error.name is always 'Error'

File: frontend/src/libs/packages/api/base-http-api.ts [hunk 6]
```diff
@@ -106,6 +106,15 @@ class BaseHttpApi implements HTTPApi {
       };
     }
 
+    if (response.status === HTTPCode.UNAUTHORIZED) {
+      throw new AuthError({
+        status: HTTPCode.UNAUTHORIZED,
+        errorType: ServerErrorType.AUTHORIZATION,
+        message: parsedException.message,
+        details: 'details' in parsedException ? parsedException.details : [],
```
Comment (v0ldemar01) [inline frontend/src/libs/packages/api/base-http-api.ts]: Not sure whether we need errorType, details and status

File: frontend/src/libs/packages/http/libs/exceptions/auth-error.exception.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+type Constructor = {
+  status: ValueOf<typeof HTTPCode>;
+  details: ServerErrorDetail[];
+  message: string;
+  errorType: ValueOf<typeof ServerErrorType>;
+  cause?: unknown;
+};
+
+class AuthError extends LibraryHTTPError {
```
Comment (v0ldemar01) [inline frontend/src/libs/packages/http/libs/exceptions/auth-error.exception.ts]: Extend just from Error

**Review by v0ldemar01 — CHANGES_REQUESTED** [review]

File: frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts (hunk 4 above)
Comment (v0ldemar01) [reply frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts]: Just pass name to custom error

File: frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts (hunk 5 above)
Comment (OlesiaTr) [reply frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts]: Done!

File: frontend/src/libs/packages/http/libs/exceptions/auth-error.exception.ts (hunk 7 above)
Comment (OlesiaTr) [reply frontend/src/libs/packages/http/libs/exceptions/auth-error.exception.ts]: Changed that!

File: frontend/src/libs/packages/api/base-http-api.ts (hunk 6 above)
Comment (OlesiaTr) [reply frontend/src/libs/packages/api/base-http-api.ts]: Deleted those!

File: frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts (hunk 4 above)
Comment (OlesiaTr) [reply frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts]: Done!

File: frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+import { ServerErrorType } from '#libs/enums/enums.js';
+import { type AppDispatch } from '#libs/packages/store/libs/types/types.js';
+import { actions as authActions } from '#slices/auth/auth.js';
+
+const handleUnauthorized: Middleware<unknown, unknown, AppDispatch> = ({
+  dispatch,
+}) => {
+  return (next) => {
+    return (action: AnyAction) => {
+      if (isRejected(action)) {
```
Comment (v0ldemar01) [inline frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts]: I believe that isRejected(action) is useless bcos action.error.name === ServerErrorType.AUTHORIZATION will kill two birds with one stone the defendant

**Review by v0ldemar01 — CHANGES_REQUESTED** [review]

**Review by HIMDanny — DISMISSED** [review]

File: frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts (hunk 8 above)
Comment (OlesiaTr) [reply frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts]: @v0ldemar01  without isRejected just on the page load I get this
isRejected

And I am basically stuck on the sign in page after it

Comment (liza-veis) [reply frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts]: we should leave isRejected

**Review by liza-veis — DISMISSED** [review]

File: frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts (hunk 8 above)
Comment (v0ldemar01) [reply frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts]: @OlesiaTr but action.error.name === ServerErrorType.AUTHORIZATION is only for 401 case and it includes rejected action

**Review by v0ldemar01 — CHANGES_REQUESTED** [review]

File: frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import { type AnyAction, type Middleware } from '@reduxjs/toolkit';
+import { isRejected } from '@reduxjs/toolkit';
```
Comment (fshabanov) [inline frontend/src/libs/packages/store/middlewares/handle-unauthorized/handle-unauthorized.middleware.ts]: Combine imports

**Review by fshabanov — CHANGES_REQUESTED** [review]

**Review by v0ldemar01 — APPROVED** [review]

**Pull request merged**